### PR TITLE
Add advancement display API

### LINKED ISSUE #6382: API Addition: FrameType
### Is your feature request related to a problem?

The default Advancement message looks ugly, so I would like to create my own by listening to the PlayerAdvancementDoneEvent event and reconstructing the message.

I will need to make my own translatable component for the advancement names. For these, I would need to get the “Frame Name”. However, the FrameType needed is not exposed to the API.

### Describe the solution you'd like.

The FrameType needs to be exposed to the API.

### Describe alternatives you've considered.

Doing some NMS magic myself, though that would be a lot more complicated.

### Agreements

- [x] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

_No response_

### PR DESCRIPTION
**Pull request by syldium**
Adds API for displaying advancements.
This allows custom menus and the use of translatable components.
Example: https://gist.github.com/syldium/4f304340ee26d31bea009295a7196522

### CODE REVIEW

File: patches/server/0727-Add-advancement-display-API.patch [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
++        return switch (this.handle.getFrame()) {
++            case TASK -> Frame.TASK;
++            case CHALLENGE -> Frame.CHALLENGE;
++            case GOAL -> Frame.GOAL;
++        };
++    }
++
++    @Override
++    public @NotNull Component title() {
++        return PaperAdventure.asAdventure(this.handle.getTitle());
```
Comment (Machine-Maker) [inline patches/server/0727-Add-advancement-display-API.patch]: Perhaps both this and the description should be cached? I’m unsure so this isn’t a needed change for the moment.

**Review by Machine-Maker — COMMENTED** [review]
The whole `AdvancementTree` interface seems odd to me, why weren't the methods on that interface just added to the regular Advancement interface and implemented in CraftAdvancement?

**Comment by syldium** [comment]
> The whole `AdvancementTree` interface seems odd to me, why weren't the methods on that interface just added to the regular Advancement interface and implemented in CraftAdvancement?

The idea is that the background property has an effect only on the root advancement when the tab is created. The `AdvancementTree` interface suggests that this property is more related to the whole tree rather than to a single advancement for me.

**Comment by Machine-Maker** [comment]
Yeah, I understand that, but I feel like it's weird to deviate from the nms DisplayInfo class so much which just has a nullable ResourceLocation for that background.

**Comment by syldium** [comment]
`AdvancementTree` interface removed to match NMS, `CraftItemStack#asCraftMirror` → `CraftItemStack#asBukkitCopy` in `AdvancementDisplay#icon`.

File: patches/server/0728-Add-advancement-display-API.patch [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+--- a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
++++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
+@@ -27,4 +27,38 @@ public class CraftAdvancement implements org.bukkit.advancement.Advancement {
+     public Collection<String> getCriteria() {
+         return Collections.unmodifiableCollection(this.handle.getCriteria().keySet());
+     }
++    // Paper start
++    @Override
++    public io.papermc.paper.advancement.AdvancementDisplay getDisplay() {
++        return this.handle.getDisplay() == null ? null : new io.papermc.paper.advancement.PaperAdvancementDisplay(this.handle.getDisplay());
```
Comment (Machine-Maker) [inline patches/server/0728-Add-advancement-display-API.patch]: Should a new instance be created for each getDisplay call?

*just a comment, not a suggestion, need for feedback to determine whats better*

**Review by Machine-Maker — COMMENTED** [review]
I'm not sure ComponentLike is appropriate for Advancement? Some adventure peeps should weigh in here.

**Comment by kezz** [comment]
> I'm not sure ComponentLike is appropriate for Advancement? Some adventure peeps should weigh in here.

I think just `Translatable` is fine. Similarly `Frame` doesn't really need to implement `StyleBuilderApplicable` (as is the case with the rest of Paper/etc).

**Comment by syldium** [comment]
Implements #6382.
I don't think that `Advancement` should implement `Translatable`, because datapacks can create new ones without translatable components. `#asComponent` exposes the name wrapped in square brackets used in the chat.

File: patches/server/0732-Add-advancement-display-API.patch [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+--- a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
++++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
+@@ -27,4 +27,38 @@ public class CraftAdvancement implements org.bukkit.advancement.Advancement {
+     public Collection<String> getCriteria() {
+         return Collections.unmodifiableCollection(this.handle.getCriteria().keySet());
+     }
++    // Paper start
++    @Override
++    public io.papermc.paper.advancement.AdvancementDisplay getDisplay() {
++        return this.handle.getDisplay() == null ? null : new io.papermc.paper.advancement.PaperAdvancementDisplay(this.handle.getDisplay());
```
Comment (Machine-Maker) [inline patches/server/0732-Add-advancement-display-API.patch]: This shouldn't be re-created every time getDisplay is called. Add a field to CraftAdvancement and set it either in the first getDisplay call, or the ctor.

File: patches/server/0732-Add-advancement-display-API.patch [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
++import org.bukkit.craftbukkit.util.CraftNamespacedKey;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public record PaperAdvancementDisplay(DisplayInfo handle) implements AdvancementDisplay {
++
++    @Override
++    public @NotNull Frame frame() {
++        return switch (this.handle.getFrame()) {
```
Comment (Machine-Maker) [inline patches/server/0732-Add-advancement-display-API.patch]: To facilitate the unit testing of bukkit `Frame` and nms `FrameType` this should probably be moved to static method to convert from nms to bukkit.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Needs unit tests to check the Frame translationKey and the Frame NamedTextColor are equal to their NMS counterparts.

EDIT: Also, Advancement shouldn't implement ComponentLike

File: patches/api/0329-Add-advancement-display-API.patch [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
++
++        /**
++         * "Advancement made" advancement.
++         */
++        TASK("task", NamedTextColor.GREEN);
++
++        /**
++         * The name map.
++         */
++        public static final Index<String, Frame> NAMES = Index.create(Frame.class, color -> color.name);
```
Comment (kezz) [inline patches/api/0329-Add-advancement-display-API.patch]:
```suggestion
+        public static final Index<String, Frame> NAMES = Index.create(Frame.class, frame -> frame.name);
```

**Review by Machine-Maker — APPROVED** [review]

**Review by JRoy — APPROVED** [review]

**Review by kashike — APPROVED** [review]

**Pull request merged**